Engine: IndexOf add string index, count and case sensitivity params

ericoporto opened the pull request:
fix #2487

- I don't remember how to deal with the previous script API being only 1 parameter in ags4 branch! (Should I also link to  `String::IndexOf^1` and then have both be able to set through ifdefs in the header? Or is this not needed in ags4 ? the defaults will keep them working but rebuild is still needed.
- I am not convinced that the final code is actually getting proper offset in unicode characters, even as is in the current branch, did we test this at some point?
- I stole bits of the substring function to make the substring that the IndexOf will apply to. I wasn't sure if I should error here too or just throw a warn and return -1.

ivan-mogilko commented:
>  (Should I also link to String::IndexOf^1 and then have both be able to set through ifdefs in the header? Or is this not needed in ags4

I am not sure myself, but if need arises then this will be trivial to add another table entry.

ivan-mogilko commented on Engine/ac/string.cpp:
```diff
@@ -427,30 +427,55 @@ int StringToInt(const char*stino) {
+    size_t byte_len = end - start;
+    size_t u_len = endIndex - startIndex;
+
+    auto buf = ScriptString::CreateBuffer(byte_len, u_len);
```
There's virtually no reason to use ScriptString::CreateBuffer here, this function is only ever needed when you need to create and return a new ScriptString object. This function creates a buffer with a meta header prepended to it.
Here you just run a standard string algorithm over a pair of buffers, so ags_strdup is correct.

ivan-mogilko replied:
Alright, I see, you need a substring.
In such case you could simply use a common String, which has a constructor accepting char pointer and length.
You may use String for both buffers, and it will automatically delete itself too.

ericoporto replied:
> then it may be convenient to have helper functions mimicing "strdup" but return a C++ container

I think this sounds best but wasn't sure where to put this helper. I thought about mimicking but taking (optional?) start and end parameters so it can either duplicate or substring.

ivan-mogilko replied:
> I think this sounds best but wasn't sure where to put this helper

StrUtil namespace is a good place for this, in string_utils.h/cpp

ericoporto replied:
> If using String, it will need to be added unicode-aware MakeLower/Upper methods (similar to existing String::ReverseUTF8).

Erh, just to be sure I am going the right way, the reason I thought to not go this way is because string.cpp is currently independent from allegro, so the only way to keep it this way would be to duplicate the utolower/utoupper functions that apply on a single char.

ivan-mogilko replied:
> the reason I thought to not go this way is because string.cpp is currently independent from allegro, so the only way to keep it this way would be to duplicate the utolower/utoupper functions that apply on a single char.

Hmm, right, getting allegro dependency there will be not good.
So either the raw buffer way where a substring is returned as `std::vector<char>`, or if in String then it has to be fully recreated there, using at most only Utf8 utilities header.

ericoporto replied:
I ended up adding to the StrUtil namespace.

ivan-mogilko commented:
> I am not convinced that the final code is actually getting proper offset in unicode characters, even as is in the current branch,

It looks like returning a correct result in the current branch, because it does this after finding a "needle":
```
*offs = 0;
int at = ustrlen(tempbuf1);
```
This puts a null terminator right where the "needle" was found, and calculates the ulength of the string before that.

You can do the same. Or you may try to optimize this, and make a separate version of ustrstr that also has a pointer to "index" as an argument, and fills in an index at which a substring was found.

ericoporto commented:
Right, this reminds me, stupid question, but the returned index is relative to the 0 character of the original string right? Not the substring.

ivan-mogilko commented:
> Right, this reminds me, stupid question, but the returned index is relative to the 0 character of the original string right? Not the substring.

Yes, it should be an index in the full string; the input parameters are also indexes in full string, so return value is consistent with them.

ericoporto commented:
ok, I think I fixed all things. I also ended adding a table entry for `IndexOf^1` pointing to `StrContains` as it was.

ericoporto commented:
ok, I think I fixed all things. I also ended adding a table entry for `IndexOf^1` pointing to `StrContains` as it was.

Also I had a [copy paste error](https://github.com/adventuregamestudio/ags/compare/9e2c6f4bf75f828d3764216187279af3bd99c61b..c6568fb44d64a37dee9201e663a9e578b1bd91a1) that is why it wasn't working at all, fixed now.

ivan-mogilko commented:
We must clarify what does "end index" mean in IndexOf: is it a *last* character of a searched range, or a first index *after* range. Right now the function works as if it were the index *after*. In other words startIndex is inclusive and endIndex is exclusive, the search is done in range `[startIndex, endIndex)`. Is that a expected behavior?

In regards to the code, I'd argue that for the new Substring function it may be better to have start and length parameters, that's common for "substring" functions.

ericoporto commented:
I think Length is less ambiguous than end index, wouldn't it be better to use it instead as input parameter?

ivan-mogilko commented:
Personally I don't mind "end index" so long as it's clearly documented; and in C and C++ it's traditionally used as an exclusive range ending (i.e. how iterator parameters work).

@messengerbag, you have suggested this parameter, what do you think about this?

ericoporto commented:
I think C# uses a "count" which would be length in the [C# String.IndexOf](https://learn.microsoft.com/en-us/dotnet/api/system.string.indexof?view=net-8.0). There are a lot of overloads, so not so sure if there's one that would also use index instead.

ivan-mogilko commented:
Alright, we may use "count" too. It's easy to convert to count if you know ending index.

ericoporto commented:
> BTW in C#'s example the "comparison type" is the last argument after startIndex and count.

Uhm, we also have it as last parameter in the other String methods we have, I think switching it here too for consistency may be a good idea. Do you agree?

ivan-mogilko commented:
Maybe call this parameter "count"? Count is number of characters after index, length is a length of an object, like a string.

ivan-mogilko commented:
Maybe call this parameter "count"? Count is number of characters after index, length is a length of an object, like a string.

> we also have it as last parameter in the other String methods we have, I think switching it here too for consistency may be a good idea. Do you agree?

Yes, I think that will be alright.

ericoporto commented:
ok, just a check, if we are inverting and switching for count, I think 0 would be better than -1 to be evaluated as instead going through the full length. I think that would be easier to remember as default for it.

ivan-mogilko commented:
> if we are inverting and switching for count, I think 0 would be better than -1 to be evaluated as instead going through the full length.

That makes sense, because 0 is an invalid value for this operation.

ericoporto commented:
ok, I think now all is correct!

messengerbag commented:
> @messengerbag, you have suggested this parameter, what do you think about this?

I suppose this has already been settled, but since I was asked, I personally find that endIndex is usually more convenient, and I think making it non-inclusive would be preferable. That is the way which, in my experience, most often allows you to simply use the variable value you already have, without any kind of arithmetic.

But obviously, the conversion between the alternatives is pretty trivial, so it's not a huge deal.

ericoporto commented:
Yeah, I had made endIndex non inclusive and also because in my head it was more convenient but in the end I think using count is easier to understand, less ambiguous, plus I think it will be easier to figure it out that 0 causes it to run until string end.

I am trying to make the API easier to get by reading parameter names and the tooltip comment and I think this achieved design with count does make sense in a somewhat more intuitive way.
